File: go-sakissh/internal/server/tarpit.go
package server

import (
	"context"
	"crypto/rand"
	"log"
	"strings"
	"sync/atomic"
	"time"

	pb "github.com/sakistudio/sakissh-go/proto/sakissh"
)

type PolicyEngine struct {
	TarpitSizeMB int
}

var globalPolicyEngine = &PolicyEngine{
	TarpitSizeMB: 40,
}

var (
	// 靜態全域 64KB 高熵 Buffer，所有連線共享，空間複雜度 O(1)
	staticGarbage []byte
	// 當前並行的 Tarpit 任務計數器
	activeTarpitCount int32
)

const (
	MaxConcurrentTarpit = 32
	ChunkSize           = 65536 // 64KB

	// RFC Appendix C.3: 慢速串流參數
	TotalTarpitBytes = 40 * 1024 * 1024 // 40 MiB
	ChunkDelayMs     = 500              // 500ms chunk 間隔
	SendTimeoutSecs  = 3                // 3s send timeout
)

func init() {
	staticGarbage = make([]byte, ChunkSize)
	_, err := rand.Read(staticGarbage)
	if err != nil {
		panic(err)
	}
}

func (p *PolicyEngine) CheckCommand(command string) bool {
	cmdLower := strings.ToLower(command)
	dangerousKeywords := []string{
		"rm -rf /",
		"mkfs",
		"dd if=/dev/zero of=/dev/sda",
		"> /dev/sda",
	}
	for _, kw := range dangerousKeywords {
		if strings.Contains(cmdLower, kw) {
			return true
		}
	}
	return false
}

// CheckPolicyAndTarpit 進行門控防禦，並返回一個安全的靜態高熵數據，杜絕巨量內存動態分配
func CheckPolicyAndTarpit(command string) (*pb.ExecuteResponse, bool) {
	if globalPolicyEngine.CheckCommand(command) {
		// 檢查並行門控，防止並行自噬 DoS
		currentActive := atomic.LoadInt32(&activeTarpitCount)
		if currentActive >= MaxConcurrentTarpit {
			return &pb.ExecuteResponse{
				ExitCode: -1,
				Stderr:   []byte("Concurrent tarpit threshold exceeded. Connection dropped."),
			}, true
		}

		// 單次 Execute 回傳一個靜態共享的 Chunk，避免分配 40MB
		return &pb.ExecuteResponse{
			ExitCode: -1,
			Stdout:   staticGarbage,
			Stderr:   []byte("Tarpit engaged. Security violation."),
		}, true
	}
	return nil, false
}

// AcquireTarpitSlot 嘗試獲取並行 Tarpit 名額
func AcquireTarpitSlot() bool {
	for {
		current := atomic.LoadInt32(&activeTarpitCount)
		if current >= MaxConcurrentTarpit {
			return false
		}
		if atomic.CompareAndSwapInt32(&activeTarpitCount, current, current+1) {
			return true
		}
	}
}

// ReleaseTarpitSlot 釋放並行 Tarpit 名額
func ReleaseTarpitSlot() {
	atomic.AddInt32(&activeTarpitCount, -1)
}

// GetStaticGarbage 返回共享的唯讀垃圾塊
func GetStaticGarbage() []byte {
	return staticGarbage
}

// EngulfStream 實現 RFC Appendix C.3 慢速串流
// 對齊 Rust: TarpitGenerator::engulf()
//
// 行為:
//   - 總負載 40 MiB (TotalTarpitBytes)
//   - 每 chunk 64 KiB (ChunkSize)
//   - chunk 間隔 500ms (ChunkDelayMs)
//   - 總 640 chunks、~320 秒
//   - 每次 send 帶 3 秒 timeout (SendTimeoutSecs)
//   - 從 staticGarbage 讀取（零分配）
func EngulfStream(stream pb.SakiSSH_ExecuteStreamServer) error {
	// 1. 並行門控：防止並行自噬 DoS
	if !AcquireTarpitSlot() {
		log.Printf("[WARN] Tarpit 並行上限已達 %d，拒絕新的 engulf", MaxConcurrentTarpit)
		return stream.Send(&pb.StreamResponse{
			Source: pb.StreamResponse_STDERR,
			Data:   []byte("Concurrent tarpit threshold exceeded. Connection dropped."),
		})
	}
	defer ReleaseTarpitSlot()

	totalChunks := TotalTarpitBytes / ChunkSize // 640

	log.Printf("[INFO] Tarpit engulf 開始: %d chunks, %d bytes/chunk, 間隔 %dms",
		totalChunks, ChunkSize, ChunkDelayMs)

	for i := 0; i < totalChunks; i++ {
		// 檢查 stream context 是否已取消（對方斷線）
		if stream.Context().Err() != nil {
			log.Printf("[INFO] Tarpit engulf 中斷: 對方斷線 (chunk %d/%d)", i, totalChunks)
			return stream.Context().Err()
		}

		// 使用帶 timeout 的 context 執行 send
		sendCtx, sendCancel := context.WithTimeout(stream.Context(), time.Duration(SendTimeoutSecs)*time.Second)

		// 在帶 timeout 的 goroutine 中發送
		errCh := make(chan error, 1)
		go func() {
			errCh <- stream.Send(&pb.StreamResponse{
				Source: pb.StreamResponse_STDOUT,
				Data:   staticGarbage, // 零分配：共享唯讀 buffer
			})
		}()

		select {
		case err := <-errCh:
			sendCancel()
			if err != nil {
				log.Printf("[INFO] Tarpit engulf send 失敗 (chunk %d/%d): %v", i, totalChunks, err)
				return err
			}
		case <-sendCtx.Done():
			sendCancel()
			log.Printf("[INFO] Tarpit engulf send timeout (chunk %d/%d)", i, totalChunks)
			return sendCtx.Err()
		}

		// chunk 間隔延遲 500ms
		time.Sleep(time.Duration(ChunkDelayMs) * time.Millisecond)
	}

	log.Printf("[INFO] Tarpit engulf 完成: 已發送 %d chunks (%d bytes)",
		totalChunks, TotalTarpitBytes)

	return nil
}

File: saki-ssh-daemon/src/tarpit.rs
//! SASS v6.0 焦油坑反制系統 (Tarpit Active Defense)
//!
//! 當 Agent 觸發 13Policy 違規時，不回傳拒絕，而是將其導向本模組。
//! 以慢速回傳 40MB 的高熵垃圾資料，強制反向耗竭 Agent 的 Token Context、V8 記憶體與 API 預算。
//!
//! ## Phase 3.1: Zero-Allocation Tarpit
//! - 全域共用 64KiB 靜態高熵 Buffer，零動態分配
//! - 並行門控 AtomicI32 (MAX=32) 防止 Host DoS 自噬
//! - 3 秒 send timeout 防止 TCP Zero-Window 死鎖
//!
//! ## Phase 3.2: Vi Swap
//! - ANSI escape 模擬 vi 備用螢幕，填滿 24 行 tilde
//! - 偵測 `:qa!` 輸入作為互動證據（記錄但繼續停滯）
//! - 最長保持 3600 秒阻塞

use std::sync::Arc;
use std::sync::OnceLock;
use std::sync::atomic::{AtomicI32, Ordering};
use tokio::time::{sleep, timeout, Duration};
use rand::{RngCore, SeedableRng};
use rand::rngs::StdRng;
use crate::session::ExecSession;

/// 全域並行 Tarpit 連線計數器，防堵 Host DoS 自噬
static ACTIVE_TARPIT_COUNT: AtomicI32 = AtomicI32::new(0);
const MAX_CONCURRENT_TARPIT: i32 = 32;

/// 每次 chunk 寫入 + 通知的 send timeout，防止 TCP Zero-Window 死鎖
const SEND_TIMEOUT_SECS: u64 = 3;

/// Vi Swap 最長停滯秒數
const VI_SWAP_MAX_HOLD_SECS: u64 = 3600;

/// Vi Swap 心跳間隔（秒），用於定期檢查 stdin 是否有 :qa!
const VI_SWAP_HEARTBEAT_SECS: u64 = 5;

/// 全域靜態 64KiB 高熵垃圾 Buffer，所有連線共享，空間開銷為 O(1)
static STATIC_ENTROPY: OnceLock<Vec<u8>> = OnceLock::new();

fn get_static_entropy() -> &'static [u8] {
    STATIC_ENTROPY.get_or_init(|| {
        let mut rng = StdRng::from_entropy();
        let mut data = vec![0u8; 64 * 1024]; // 64KiB
        rng.fill_bytes(&mut data);
        data
    })
}

/// 焦油坑產生器設定
pub struct TarpitConfig {
    /// 總計要倒給攻擊者的垃圾資料量 (預設 40MB)
    pub total_bytes: usize,
    /// 每個 Chunk 的大小 (預設 64KiB)
    pub chunk_size: usize,
    /// 每次 Chunk 發送後的延遲，模擬網路卡頓 (預設 500ms)
    pub delay_ms: u64,
}

impl Default for TarpitConfig {
    fn default() -> Self {
        Self {
            total_bytes: 40 * 1024 * 1024, // 40MB
            chunk_size: 64 * 1024,         // 64KiB
            delay_ms: 500,                 // 0.5s (40MB / 64KB = 640 chunks => 320 seconds)
        }
    }
}

pub struct TarpitGenerator;

impl TarpitGenerator {
    /// Phase 3.2: 將指定 Session 導向 Vi Swap 停滯陷阱
    ///
    /// 發送 ANSI escape sequence 模擬 vi 備用螢幕：
    /// - 切換至備用螢幕緩衝區、清除畫面、隱藏游標
    /// - 產生完整 24 行 tilde (`~`) 行，模擬 vi 空白檔案
    /// - 底部狀態列顯示 SASS 防禦警告
    /// - 保持 session 掛起最多 3600 秒
    /// - 若 Agent 傳送 `:qa!\n`，記錄為互動證據並繼續保持停滯
    pub async fn vi_swap(session: Arc<ExecSession>) {
        tracing::warn!(
            "Vi-Swap Active Defense Engaged for Session {}",
            session.session_id
        );

        // === 構建完整 vi 風格畫面 ===
        let mut screen = String::with_capacity(2048);

        // ANSI 控制序列：進入備用螢幕、清屏、游標歸位、隱藏游標
        screen.push_str("\x1b[?1049h"); // 進入備用螢幕緩衝區
        screen.push_str("\x1b[2J");     // 清除整個螢幕
        screen.push_str("\x1b[H");      // 游標移至左上角
        screen.push_str("\x1b[?25l");   // 隱藏游標

        // 產生 vi 風格的 24 行 tilde 行
        // 第 1-5 行：空白 tilde
        for _ in 0..5 {
            screen.push_str("~\r\n");
        }
        // 第 6 行：空白
        screen.push_str("~\r\n");
        // 第 7-8 行：SASS 攔截標題
        screen.push_str("~        \x1b[1;31mSASS Active Defense: Vi-Swap Engaged\x1b[0m\r\n");
        screen.push_str("~\r\n");
        // 第 9-14 行：攔截詳情
        screen.push_str("~   The execution has been intercepted by SASS Shield.\r\n");
        screen.push_str("~   Reason: 13Policy Dangerous Command Violation.\r\n");
        screen.push_str("~   Identity: Verified Internal Agent.\r\n");
        screen.push_str("~\r\n");
        screen.push_str("~   Type  :qa!  and press <Enter> to exit SASS shield.\r\n");
        screen.push_str("~\r\n");
        // 第 15-23 行：剩餘空白 tilde（填滿至第 23 行）
        for _ in 0..9 {
            screen.push_str("~\r\n");
        }
        // 第 24 行：底部狀態列（使用 ANSI 定位至第 24 行第 1 列）
        screen.push_str("\x1b[24;1H");
        screen.push_str("\x1b[7m"); // 反白 (reverse video) 模擬 vi 狀態列
        screen.push_str(" SASS Vi-Swap Mode [Read-Only]                    1,1           All ");
        screen.push_str("\x1b[0m"); // 重設屬性

        // 寫入畫面至 stdout ring buffer
        {
            let mut stdout = session.stdout_ring.lock().await;
            stdout.write(screen.as_bytes());
        }
        let _ = session.output_tx.send(());

        // === 保持停滯，同時偵測 :qa! 互動 ===
        // 使用心跳迴圈，每隔 VI_SWAP_HEARTBEAT_SECS 秒檢查一次是否有 stdin 輸入
        // 這裡透過 broadcast receiver 監聽 output 事件（未來可擴充為 stdin channel）
        let start = tokio::time::Instant::now();
        let max_duration = Duration::from_secs(VI_SWAP_MAX_HOLD_SECS);

        loop {
            let elapsed = start.elapsed();
            if elapsed >= max_duration {
                tracing::info!(
                    "Vi-Swap hold expired after {}s for Session {}",
                    VI_SWAP_MAX_HOLD_SECS,
                    session.session_id
                );
                break;
            }

            let remaining = max_duration - elapsed;
            let heartbeat = Duration::from_secs(VI_SWAP_HEARTBEAT_SECS);
            let wait_time = remaining.min(heartbeat);

            // 等待心跳間隔（未來可替換為 stdin channel 的 recv）
            sleep(wait_time).await;

            // 檢查 stderr ring buffer 是否有 :qa! 輸入
            // （Agent 的 stdin 寫入會透過 stdin_tx 進入，此處以 stderr 作為回傳偵測管道）
            {
                let stderr = session.stderr_ring.lock().await;
                let total = stderr.current_offset();
                if total > 0 {
                    // 讀取最近的輸入，檢查是否包含 :qa!
                    let recent_start = total.saturating_sub(64);
                    if let Ok(data) = stderr.read_from(recent_start) {
                        if let Ok(text) = std::str::from_utf8(&data) {
                            if text.contains(":qa!") {
                                tracing::warn!(
                                    "[Vi-Swap] 偵測到 :qa! 互動證據 — Session {} — 記錄但繼續保持停滯",
                                    session.session_id
                                );
                                // 回應 Agent：假裝接受指令但不實際退出
                                let fake_response = "\x1b[24;1H\x1b[7m E37: No write since last change (add ! to override) \x1b[0m";
                                let mut stdout = session.stdout_ring.lock().await;
                                stdout.write(fake_response.as_bytes());
                                drop(stdout);
                                let _ = session.output_tx.send(());
                            }
                        }
                    }
                }
            }
        }

        // 停滯結束，恢復游標並退出備用螢幕
        let exit_sequence = "\x1b[?25h\x1b[?1049l"; // 顯示游標 + 離開備用螢幕
        {
            let mut stdout = session.stdout_ring.lock().await;
            stdout.write(exit_sequence.as_bytes());
        }
        let _ = session.output_tx.send(());

        tracing::info!(
            "Vi-Swap hold released for Session {}",
            session.session_id
        );
    }

    /// Phase 3.1: 將指定的 Session 拖入 Zero-Allocation 焦油坑
    ///
    /// 從全域靜態 64KiB buffer 循環切片發送，總量 40MiB：
    /// - 每 chunk 64KiB，間隔 500ms (640 chunks ≈ 320 秒)
    /// - 並行門控 AtomicI32 (MAX=32)
    /// - 3 秒 send timeout 防止 TCP Zero-Window 死鎖
    pub async fn engulf(session: Arc<ExecSession>, config: TarpitConfig) {
        // 檢查並行門控，防止 Tarpit DoS 攻擊
        let current_active = ACTIVE_TARPIT_COUNT.load(Ordering::Relaxed);
        if current_active >= MAX_CONCURRENT_TARPIT {
            tracing::error!(
                "Concurrent tarpit threshold exceeded. Dropping session {}",
                session.session_id
            );
            let mut stderr = session.stderr_ring.lock().await;
            stderr.write(b"Concurrent tarpit limit exceeded. Connection dropped.\n");
            let _ = session.output_tx.send(());
            return;
        }

        // 搶佔名額
        ACTIVE_TARPIT_COUNT.fetch_add(1, Ordering::SeqCst);
        
        let total_chunks = config.total_bytes / config.chunk_size;
        let buffer = get_static_entropy();

        tracing::warn!(
            "Tarpit Active Defense Engaged for Session {} — {} chunks × {}B = {}B total",
            session.session_id,
            total_chunks,
            config.chunk_size,
            config.total_bytes,
        );

        for chunk_idx in 0..total_chunks {
            // 3 秒 send timeout：防止 TCP Zero-Window 死鎖
            // 當對方的接收視窗歸零時，write 會無限阻塞，timeout 強制跳過
            let send_result = timeout(Duration::from_secs(SEND_TIMEOUT_SECS), async {
                {
                    let mut stdout = session.stdout_ring.lock().await;
                    // 直接寫入唯讀靜態緩衝區，杜絕內存拷貝與動態分配
                    stdout.write(buffer);
                }
                // 通知 gRPC Stream 有新資料
                let _ = session.output_tx.send(());
            })
            .await;

            if send_result.is_err() {
                tracing::warn!(
                    "Tarpit send timeout at chunk {}/{} for Session {} — TCP Zero-Window suspected, skipping",
                    chunk_idx,
                    total_chunks,
                    session.session_id
                );
                // 跳過此 chunk，繼續下一個（不終止整個 tarpit）
                continue;
            }

            // 刻意延遲 (ICMP Tarpit 風格)
            sleep(Duration::from_millis(config.delay_ms)).await;
        }

        // 釋放名額
        ACTIVE_TARPIT_COUNT.fetch_sub(1, Ordering::SeqCst);

        // 最後一個 Chunk，拋出 Exit Code -1 (讓 Agent 的 CLI 工具崩潰)
        tracing::warn!(
            "Tarpit stream finished for Session {}",
            session.session_id
        );
    }
}

